src/app/public/modules/colorpicker/colorpicker.component.spec.ts

```diff
@@ -186,443 +211,640 @@ describe('Colorpicker Component', () => {
-    setInputElementValue(nativeElement, 'hex', '#654321');
-    verifyColorpicker(nativeElement, 'cmyk(0%,34%,67%,60%)', '101, 67, 33');
-  }));
-
-  it('should accept transparency', fakeAsync(() => {
-    component.selectedOutputFormat = 'hsla';
-    openColorpicker(nativeElement, fixture);
-    setInputElementValue(nativeElement, 'red', '0');
-    setInputElementValue(nativeElement, 'green', '0');
-    setInputElementValue(nativeElement, 'blue', '0');
-    setInputElementValue(nativeElement, 'alpha', '0');
-    verifyColorpicker(nativeElement, 'hsla(0,0%,0%,0)', '0, 0, 0, 0');
-  }));
-
-  it('should accept color change through directive host listener', fakeAsync(() => {
-    component.selectedOutputFormat = 'rgba';
-    openColorpicker(nativeElement, fixture);
-    nativeElement.querySelector('input').value = '#4523FC';
-    let inputEvent = document.createEvent('Event');
-    inputEvent.initEvent('input', true, false);
-    let changeEvent = document.createEvent('Event');
-    changeEvent.initEvent('change', true, false);
-    nativeElement.querySelector('input').dispatchEvent(inputEvent);
-    nativeElement.querySelector('input').dispatchEvent(changeEvent);
-    fixture.detectChanges();
-    verifyColorpicker(nativeElement, 'rgba(69,35,252,1)', '69, 35, 252');
-  }));
-
-  it('should allow user to esc cancel the color change.', fakeAsync(() => {
-    component.selectedOutputFormat = 'hex';
-    openColorpicker(nativeElement, fixture);
-    setInputElementValue(nativeElement, 'hex', '#086A93');
-    verifyColorpicker(nativeElement, '#086a93', '8, 106, 147');
-    keyHelper('Escape', 27, 'Esc');
-    fixture.detectChanges();
-    verifyColorpicker(nativeElement, '#2889e5', '40, 137, 229');
-  }));
-
-  it('should specify type="button" on all button elements.', fakeAsync(() => {
-    component.selectedOutputFormat = 'hex';
-    openColorpicker(nativeElement, fixture);
-    expect(nativeElement.querySelectorAll('button:not([type="button"])').length).toBe(0);
-  }));
+    beforeEach(() => {
+      fixture = TestBed.createComponent(ColorpickerTestComponent);
+      nativeElement = fixture.nativeElement as HTMLElement;
+      component = <ColorpickerTestComponent> fixture.componentInstance;
+      colorpickerComponent = component.colorpickerComponent;
+    });
 
-  it('should hide when input type is set to hidden.', fakeAsync(() => {
-    component.inputType = 'hidden';
-    const directiveEl = fixture.debugElement.query(By.directive(SkyColorpickerInputDirective));
-    const directiveInstance = directiveEl.injector.get(SkyColorpickerInputDirective);
-    tick();
-    fixture.detectChanges();
-    directiveInstance.ngOnInit();
-    tick();
-    fixture.detectChanges();
-    tick();
-    openColorpicker(nativeElement, fixture);
-    fixture.detectChanges();
-    expect(nativeElement.querySelectorAll('.sky-colorpicker-hidden').length).toEqual(1);
-  }));
+    it('should populate correct information if model is given', fakeAsync(() => {
```

Comment:
The number of changes in this spec is a tad unsettling. Can you give me some color around the changes? If it's purely for cleanup, maybe we could do that in a different pull request?

I would prefer we keep the changes specific to your feature/bug fix, as it helps me be more confident that nothing else was inadvertently broken or changed.

Reply:
No. The changes are due to a bad assumption around how things were working. The `valueChanges` was firing on every mouse drag and updating the model before anything was saved. This was allowing earlier access to the new value than it should have been. The unit tests were taking advantage of this and I had to fix the tests to now work with what was decided to be the correct behavior. Nothing about the tests spirit changed but I had to make some changes to get them to work with the intended behavior.